tests/api/test_opds2.py

```diff
@@ -86,3 +101,110 @@ def test_navigation(self):
+            self._patron(), "http://example.org/token"
+        )
+
+        assert token == INVALID_CREDENTIALS
+
+
+class TestOPDS2TokenAuthenticationFulfillmentProcessor(DatabaseTest):
+    @patch("api.opds2.HTTP")
+    def test_fulfill(self, mock_http):
+        patron = self._patron()
+        patron.username = "username"
+        collection: Collection = self._collection(
+            protocol=ExternalIntegration.OPDS2_IMPORT
+        )
+        work = self._work(with_license_pool=True, collection=collection)
+        integration: ExternalIntegration = collection.create_external_integration(
+            ExternalIntegration.OPDS2_IMPORT
+        )
+        setting: ConfigurationSetting = ConfigurationSetting.for_externalintegration(
+            ExternalIntegration.TOKEN_AUTH, integration
+        )
+        setting.value = "http://example.org/token?userName={patron_id}"
+
+        ff_info = FulfillmentInfo(
+            collection,
+            "datasource",
+            "proquest",
+            "11234",
+            "http://example.org/11234/fulfill?authToken={authentication_token}",
+            None,
+            None,
+            None,
+        )
+
+        resp = Response()
+        resp.status_code = 200
+        resp.raw = io.BytesIO(b"plaintext-auth-token")
+        mock_http.get_with_timeout.return_value = resp
+
+        processor = OPDS2TokenAuthenticationFulfillmentProcessor(collection)
+
+        ff_info = processor.fulfill(patron, None, work.license_pools[0], None, ff_info)
+
+        assert mock_http.get_with_timeout.call_count == 1
+        assert (
+            mock_http.get_with_timeout.call_args[0][0]
+            == "http://example.org/token?userName=username"
+        )
+
+        assert (
+            ff_info.content_link
+            == "http://example.org/11234/fulfill?authToken=plaintext-auth-token"
+        )
+        assert ff_info.content_link_redirect == True
+
+        ## Test error case
+        # Reset the content link
+        ff_info.content_link = (
+            "http://example.org/11234/fulfill?authToken={authentication_token}"
+        )
+        # non-200 response
+        resp = Response()
+        resp.status_code = 400
+        mock_http.reset_mock()
+        mock_http.get_with_timeout.return_value = resp
+        with pytest.raises(CannotFulfill):
+            processor.fulfill(patron, None, work.license_pools[0], None, ff_info)
+
+        ## Pass through cases
+        # No templating in the url
+        ff_info.content_link = (
+            "http://example.org/11234/fulfill?authToken=authentication_token"
+        )
+        ff_info.content_link_redirect = False
+        ff_info = processor.fulfill(patron, None, work.license_pools[0], None, ff_info)
+        assert ff_info.content_link_redirect == False
+
+        # No token endpoint config
+        ff_info.content_link = (
+            "http://example.org/11234/fulfill?authToken={authentication_token}"
```

Comment:
Might be good to add a test case for a complicated URI template, something like: 
```
http://example.org/11234/fulfill{?authentication_token}
```

expands to
```
http://example.org/11234/fulfill?authentication_token=foobarbaz
```